File: mem_caesar.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define CHUNK_SIZE 1024
#define BYTES 256

int is_numeric(char *s) {
    char first = *s;
    if (first != '-' && !isdigit(first)) return 0;
    s++;
    while (*s) {
        if (!isdigit(*s)) return 0;
        s++;
    }
    return 1;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fputs("Usage: ./caesar.php N", stderr);
        return -1;
    }
    char *S = argv[1];

    if (!is_numeric(S)) {
        fputs(S, stderr);
        fputs(" should be numeric\n", stderr);
        return -1;
    }
    
    char N = atoi(S);
    int read = 0;
    int i;
    char up_bounds[2] = {'A', 'Z'};
    char low_bounds[2] = {'a', 'z'};
    char diff = 'Z' - 'A' + 1;
    char cache[BYTES];
    char buffer[CHUNK_SIZE+1];

    memset(cache, 0, BYTES);

    while ( (read = fread(buffer, sizeof(char), CHUNK_SIZE, stdin)) > 0 ) {
        i = 0;
        while ( i < read ) {
            char current = buffer[i],
                 result;
            if (!cache[current]) {
                result = current;
                if (result >= low_bounds[0] && result <= low_bounds[1]) {
                    result += N;
                    if (result > low_bounds[1]) {
                        result -= diff;
                    }
                    if (result < low_bounds[0]) {
                        result += diff;
                    }
                }

                if (result >= up_bounds[0] && result <= up_bounds[1]) {
                    result += N;
                    if (result > up_bounds[1]) {
                        result -= diff;
                    }
                    if (result < up_bounds[0]) {
                        result += diff;
                    }
                }
                cache[current] = result;
            }
            else {
                result = cache[current];
            }
            buffer[i++] = result;

        }

        buffer[i] = 0;

        fwrite(buffer, sizeof(char), read, stdout);
    }

    return 0;
}

File: mem_caesar.py
import sys

def error_die(message):
    sys.stderr.write(message)
    sys.stderr.write("\n")
    sys.exit(-1)


if len(sys.argv) < 2:
    error_die('Usage: ./caesar.php N')

N = sys.argv[1];

if not N.isdigit() and (N[0:1] != '-' or not N[1:].isdigit()) :
    error_die("{} should be numeric".format(N))

CHUNK_SIZE = 1024
N = int(N)

cache = {}
low_bounds = ['a', 'z']
up_bounds = ['A', 'Z']
diff = ord('Z') - ord('A') + 1;
ord_low_bounds = [ord('a'), ord('z')]
ord_up_bounds = [ord('A'), ord('Z')]

while True:
    input = sys.stdin.read(CHUNK_SIZE)
    if len(input) == 0:
        break
    buffer = []
    for c in input:
        r = cache.get(c)
        if r == None:
            r = c
            if c >= low_bounds[0] and c <= low_bounds[1]:
                x = ord(c) + N
                if x > ord_low_bounds[1]:
                    x -= diff;
                if x < ord_low_bounds[0]:
                    x += diff;
                r = chr(x)

            if c >= up_bounds[0] and c <= up_bounds[1]:
                x = ord(c) + N;
                if x > ord_up_bounds[1]:
                    x -= diff;
                if x < ord_up_bounds[0]:
                    x += diff;

                r = chr(x)
            cache[c] = r

        buffer.append(r)
    sys.stdout.write("".join(buffer))
